### DISCUSSION TITLE
Too many SQL queries are fired while using .execute(select(...))

### DISCUSSION THREAD

**Discussion by Laope94** [sqlalchemy/sqlalchemy#11253] [Usage Questions] [answered]
Greetings,

I am using 2.0 style as described here - https://docs.sqlalchemy.org/en/20/changelog/migration_20.html#migration-orm-usage and select produces unnecessary amount of SQL queries.

For simplification, let's say this is my query:
`result = session.execute(select(MyEntity).filter(...)).scalars().all()`

While running this line of code first query is fired (simplified again):
`SELECT my_entity.id, my_entity.col_a, my_entity.col_b, ..., my_entity.col_x
FROM my_entity
WHERE condition_a AND condition_b ...`
Where clause is based on my .filter(...)

Let's say this retrieves 5 items from DB. However, now when I am trying to iterate over the result (I am mapping ORM to pydantic DTO), new query is fired for every item looking like this:
`SELECT my_entity.id AS my_entity_id, my_entity.col_a AS my_entity_col_a, my_entity.col_b AS my_entity_col_b, ..., my_entity.col_x AS my_entity_col_x
FROM my_entity
WHERE my_entity.id = ?`
This time with primary key column in where clause and id of item is filled to wildcard. This query is actually fired on access to item, that means, it also fires e.g. when I open result and then item in debugger, I don't know how I'd describe this behavior, but it'd compare it to lazy load (except this is not joined entity, this is a standalone table without any other relations).

Now my question is - is there a way how to retrieve all items at once with first query and prevent producing new queries on access to result items? Or is there a reason why would I want to retrieve on access?

My personal toughts are that this is due to .scalars() which (if I understand documentation correctly) returns some filtering object, and I am failing to figure this out while keeping query 2.0 styled, if I leave .scalars() out and just use .all() directly then I am facing errors in pydantic's .from_orm() function.

I am using sqlite3 as DB.

**Comment by CaselIT** [comment]
Hi,

Can you provide an example of what you are trying to do? Without it suggestion are limited
This seems like you are accessing the object after you have committed the session and you have `expire_on_commit` left at true.

**Reply by Laope94** [reply]
Well, what I am doing is really only a simple select query and mapping the result into pydantic, nothing more.

```
# Base class and ORM class
class Base(DeclarativeBase):
    pass

class MyEntity(Base):
    __tablename__ = 'my_entity'

    id = Column(String, primary_key=True)
    # other columns here... 

# DB setup
db_engine = create_engine(connection string)
session_factory = sessionmaker(db_engine, expire_on_commit=False)

# query and mapping
with session_factory() as session:
        # first SQL executes here, result is List with 5 items of MyEntity type
        result = session.execute(select(MyEntity).filter(...)).scalars().all()

# mapping to pydantic, 5 more queries are executed when items of list are accessed, as described in original post
my_entity_pd_list = [MyEntityPDModel.from_orm(my_entity_orm) for my_entity_orm in result]
```

Mapping itself is not relevant though, that 5 queries are executed on any kind of access, if I place breakpoint after getting the result and inspect list items in debugger, SQL query is executed. If I iterated over the result and just printed the item, again, query is executed.

**Reply by CaselIT** [reply]
Not sure what `from_orm` does but that seems an issue with that function or something else you omitted
There a runnable example, that show no additional queries:
```py
from sqlalchemy.engine.create import create_engine
from sqlalchemy.orm import DeclarativeBase, sessionmaker
from sqlalchemy.sql.schema import Column
from sqlalchemy import select
from sqlalchemy.sql.sqltypes import String

# Base class and ORM class
class Base(DeclarativeBase):
    pass

class MyEntity(Base):
    __tablename__ = "my_entity"

    id = Column(String, primary_key=True)
    text = Column(String)

# DB setup
db_engine = create_engine("sqlite://", echo=True)
Base.metadata.create_all(db_engine)
session_factory = sessionmaker(db_engine, expire_on_commit=False)
with session_factory() as session:
    session.add_all(
        [
            MyEntity(id="a", text="ta"),
            MyEntity(id="b", text="tb"),
            MyEntity(id="c", text="tc"),
            MyEntity(id="d", text="td"),
        ]
    )
    session.commit()

# query and mapping
with session_factory() as session:
    # first SQL executes here, result is List with 5 items of MyEntity type
    result = (
        session.execute(select(MyEntity).filter(MyEntity.id != "b"))
        .scalars()
        .all()
    )

for e in result:
    print(f'{e.id}->{e.text}')
```
the above logs the following:
```
2024-04-10 21:25:34,027 INFO sqlalchemy.engine.Engine BEGIN (implicit)
2024-04-10 21:25:34,028 INFO sqlalchemy.engine.Engine PRAGMA main.table_info("my_entity")
2024-04-10 21:25:34,028 INFO sqlalchemy.engine.Engine [raw sql] ()
2024-04-10 21:25:34,028 INFO sqlalchemy.engine.Engine PRAGMA temp.table_info("my_entity")
2024-04-10 21:25:34,028 INFO sqlalchemy.engine.Engine [raw sql] ()
2024-04-10 21:25:34,029 INFO sqlalchemy.engine.Engine
CREATE TABLE my_entity (
        id VARCHAR NOT NULL,
        text VARCHAR,
        PRIMARY KEY (id)
)


2024-04-10 21:25:34,029 INFO sqlalchemy.engine.Engine [no key 0.00014s] ()
2024-04-10 21:25:34,029 INFO sqlalchemy.engine.Engine COMMIT
2024-04-10 21:25:34,030 INFO sqlalchemy.engine.Engine BEGIN (implicit)
2024-04-10 21:25:34,031 INFO sqlalchemy.engine.Engine INSERT INTO my_entity (id, text) VALUES (?, ?)
2024-04-10 21:25:34,031 INFO sqlalchemy.engine.Engine [generated in 0.00015s] [('a', 'ta'), ('b', 'tb'), ('c', 'tc'), ('d', 'td')]
2024-04-10 21:25:34,031 INFO sqlalchemy.engine.Engine COMMIT
2024-04-10 21:25:34,032 INFO sqlalchemy.engine.Engine BEGIN (implicit)
2024-04-10 21:25:34,033 INFO sqlalchemy.engine.Engine SELECT my_entity.id, my_entity.text
FROM my_entity
WHERE my_entity.id != ?
2024-04-10 21:25:34,033 INFO sqlalchemy.engine.Engine [generated in 0.00014s] ('b',)
2024-04-10 21:25:34,033 INFO sqlalchemy.engine.Engine ROLLBACK
a->ta
c->tc
d->td
```

**Reply by Laope94** [reply, marked as answer]
Okay, I've found the cause. I've suspected query itself, because this started after code refactor to 2.0 style. However, one more thing has been changed and it slipped my attention, I didn't suspect that at all so I simplified it in code I've provided above.

Thing is that context manager function is used for transaction management to reduce boilerplate code. It looks like this:

```
@contextmanager
def transaction(read_only: bool = True):
    session = session_factory()

    try:
         yield session
    except:
         session.rollback()
         raise
    else:
         if not read_only:
             session.commit()
         else:
              session.rollback()
```

Originally, logic was simple - rollback on exception, commit if everything went fine. However, logic has been changed and read_only variable  was added to enforce that on some places no changes to DB are desired and that flag should is one of the things that should warn everyone who tries to do so, it defaults to read only, so it needs to be explicitly changed in case of the writting operation.
Last line is issue though:

```
else:
     session.rollback()
```

Originally meant to revert any change that might have been done during session marked as read only.

Why are queries still being fired during inspecting result in debugger is mystery to me though, because at the time outer else block did not happened yet.

So I think this is solved, now I only need to take a moment and think about proper way to do this, one way is to do nothing at all instead of rollback, or just do anything that is needed while keeping session opened.. or something entirely else..

Anyway, thanks for help.

**Reply by CaselIT** [reply]
You could just close the session there. Assuming you have rollback on return set in the pool (that's the default) any transaction would be then reset there
